jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PushLayersStep.java

```diff
@@ -76,34 +75,40 @@
   public ImmutableList<AsyncStep<PushBlobStep>> call() throws ExecutionException {
     try (TimerEventDispatcher ignored =
         new TimerEventDispatcher(buildConfiguration.getEventDispatcher(), DESCRIPTION)) {
-      ImmutableList<? extends AsyncStep<? extends CachedLayer>> cachedLayer =
+      ImmutableList<? extends AsyncStep<? extends CachedLayer>> cachedLayers =
           NonBlockingSteps.get(cachedLayerStep);
 
-      // Constructs a PushBlobStep for each layer.
-      ImmutableList.Builder<AsyncStep<PushBlobStep>> pushBlobStepsBuilder = ImmutableList.builder();
-      for (AsyncStep<? extends CachedLayer> cachedLayerStep : cachedLayer) {
-        ListenableFuture<PushBlobStep> pushBlobStepFuture =
-            Futures.whenAllSucceed(cachedLayerStep.getFuture())
-                .call(() -> makePushBlobStep(cachedLayerStep), listeningExecutorService);
-        pushBlobStepsBuilder.add(() -> pushBlobStepFuture);
+      try (ProgressEventDispatcher progressEventDispatcher =
+          progressEventDispatcherFactory.create("push layers", cachedLayers.size())) {
+        // Constructs a PushBlobStep for each layer.
+        ImmutableList.Builder<AsyncStep<PushBlobStep>> pushBlobStepsBuilder =
+            ImmutableList.builder();
+        for (AsyncStep<? extends CachedLayer> cachedLayerStep : cachedLayers) {
+          ProgressEventDispatcher.Factory childProgressEventDispatcherFactory =
+              progressEventDispatcher.newChildProducer();
+          ListenableFuture<PushBlobStep> pushBlobStepFuture =
+              Futures.whenAllSucceed(cachedLayerStep.getFuture())
+                  .call(
+                      () -> makePushBlobStep(cachedLayerStep, childProgressEventDispatcherFactory),
```

Comment:
You can embed `progressEventDispatcher.newChildProducer()` here instead of having `childProgressEventDispatcherFactory`? The timing will be different, but if timing matters, but I wonder if the `try` block is doing the right thing, as I said below.

Reply:
The `newChildProducer` is necessary to be outside because it needs to decrement the `remainingAllocationUnits` of `progressEventDispatcher`, whereas the actual `create` call can be done when the `pushBlobStepFuture` actually runs.